util/apiwarnings/handler_test.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiwarnings
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/go-logr/logr/funcr"
+	. "github.com/onsi/gomega"
+)
+
+func TestDiscardMatchingHandler(t *testing.T) {
+	tests := []struct {
+		name       string
+		opts       DiscardMatchingHandlerOptions
+		code       int
+		message    string
+		wantLogged bool
+	}{
+		{
+			name:    "log, if warning does not match any expression",
+			code:    299,
+			message: "non-matching warning",
+			opts: DiscardMatchingHandlerOptions{
+				Expressions: []regexp.Regexp{},
+			},
+			wantLogged: true,
+		},
+		{
+			name:    "do not log, if warning matches at least one expression",
+			code:    299,
+			message: "matching warning",
+			opts: DiscardMatchingHandlerOptions{
+				Expressions: []regexp.Regexp{
+					*regexp.MustCompile("^matching.*"),
+				},
+			},
+			wantLogged: false,
+		},
+		{
+			name:    "do not log, if code is not 299",
+			code:    0,
+			message: "",
+			opts: DiscardMatchingHandlerOptions{
+				Expressions: []regexp.Regexp{},
+			},
+			wantLogged: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			logged := false
+			h := NewDiscardMatchingHandler(
+				funcr.New(func(_, _ string) {
+					logged = true
+				},
+					funcr.Options{},
+				),
+				tt.opts,
+			)
+			h.HandleWarningHeader(tt.code, "", tt.message)
+			g.Expect(logged).To(Equal(tt.wantLogged))
+		})
+	}
+}
+
+func TestDiscardMatchingHandler_uninitialized(t *testing.T) {
```

Comment:
Should we also test this with a log that would be logged?

Reply:
Sorry, I don't understand your question.

Let me add some context: I considered not exporting `DiscardMatchingHandler`, and exporting only the `NewDiscardMatchingHandler` constructor. If I did that, the constructor would return an unexported type, which is considered "annoying," although it's [controversial](https://github.com/golang/lint/issues/465).

However, the constructor provides the only means of initializing DiscardMatchingHandler. Exporting it means someone could use uninitialized, and this tests verifies that using it uninitialized does not cause a panic, but effectively does nothing.

(If, based on the other thread here, we remove the deduplication implementation, I think I remove the constructor, and then I can also remove this test.)

Reply:
Thx, that explanation helps.

I'm basically confused that this test tests that we don't panic.

But if the HandleWarningHeader func would be called with a message & code that actually gets to `h.Logger.Info(message)` it would panic, right?

So I guess we should have:
* a test case that calls HandleWarningHeader with parameters that would usually produce a log
* add additional handling for the Logger == nil case, so we don't panic in that case

Reply:
> So I guess we should have:
> * a test case that calls HandleWarningHeader with parameters that would usually produce a log
> * add additional handling for the Logger == nil case, so we don't panic in that case

I fixed the test, so that it does exercise `h.Logger.Info(message)`.

And it turns out that `Logger` can't be nil, so we don't need to handle that. I was confused, I thought `logr.Logger` was an interface, and so it could be nil. But it's a struct.

Reply:
Ah, I missed that as well :)